lib/hash_ring/ets.ex

```diff
@@ -0,0 +1,192 @@
+  @compile :native
+
+  @default_ring_gen_gc_delay 10_000
+  @type t :: __MODULE__
+
+  use GenServer
+
+  alias HashRing.Utils
+  alias HashRing.ETS.Config
+
+  defstruct num_replicas: 0,
+            nodes: [],
+            table: nil,
+            ring_gen: 0,
+            name: nil,
+            pending_gcs: %{}
+
+  def start_link(name, nodes \\ [], num_replicas \\ 512) do
+    GenServer.start_link(__MODULE__, {name, nodes, num_replicas}, name: name)
+  end
+
+  @spec init({atom, [binary], integer}) :: t
+  def init({name, nodes, num_replicas}) do
+    table = :ets.new(name, [
+      :protected,
+      :ordered_set,
+      {:read_concurrency, true}
+    ])
+
+    state = %__MODULE__{
+      table: table,
+      num_replicas: num_replicas,
+      nodes: nodes,
+      name: name,
+    } |> rebuild
+
+    {:ok, state}
+  end
+
+  def stop(name) do
+    GenServer.stop(name)
+  end
+
+  @spec set_nodes(atom, [binary]) :: t
+  def set_nodes(name, nodes) do
+    GenServer.call(name, {:set_nodes, nodes})
+  end
+
+  @spec add_node(atom, binary) :: {:ok, t} | :error
+  def add_node(name, node) do
+    GenServer.call(name, {:add_node, node})
+  end
+
+  @spec remove_node(atom, binary) :: {:ok, t} | :error
+  def remove_node(name, node) do
+    GenServer.call(name, {:remove_node, node})
+  end
+
+  @spec get_nodes(atom) :: {:ok, [binary]} | :error
+  def get_nodes(name) do
+    GenServer.call(name, :get_nodes)
+  end
+
+  @spec force_gc(atom, integer) :: :ok | {:error, :not_pending}
+  def force_gc(name, ring_gen) do
+    GenServer.call(name, {:force_gc, ring_gen})
+  end
+
+  @spec get_ring_gen(atom) :: {:ok, integer} | :error
+  def get_ring_gen(name) do
+    with {:ok, {_, ring_gen, _}} <- Config.get(name) do
+      {:ok, ring_gen}
+    end
+  end
+
+  @spec find_node(atom, binary | integer) :: binary | nil
+  def find_node(name, key) do
+    with {:ok, config} <- Config.get(name),
+         {_, node} <- find_next_highest_item(config, Utils.hash(key)) do
+      node
+    else
+      _ -> nil
+    end
+  end
+
+  @spec find_nodes(atom, binary | integer, integer) :: [binary]
+  def find_nodes(name, key, num) do
+    with {:ok, {_, _, num_nodes}=config} <- Config.get(name),
+         nodes <- do_find_nodes(config, min(num, num_nodes), Utils.hash(key), []) do
+      nodes
+    else
+      _ -> []
+    end
+  end
+
+  ## Private
+
+  defp do_find_nodes(_config, 0, _key_int, nodes) do
+    Enum.reverse(nodes)
+  end
+  defp do_find_nodes(config, ret, key_int, nodes) do
+    {number, node} = find_next_highest_item(config, key_int)
+    if node in nodes do
+      do_find_nodes(config, ret, number, nodes)
+    else
+      do_find_nodes(config, ret - 1, number, [node|nodes])
+    end
+  end
+
+  def handle_call({:set_nodes, nodes}, _from, %{name: name}=state) do
+    new_state = %{state | nodes: nodes} |> rebuild
+    {:reply, {:ok, name}, new_state}
+  end
+  def handle_call({:add_node, node}, _from, %{name: name, nodes: nodes}=state) do
```

Comment:
can you put spaces between these functions heads? I really dislike how everything runs together in a huge block of code. Also, you're shadowing the node() function here in the args.

Reply:
hmm. generally in the past we've grouped different cases for a function with out new lines.

Reply:
also, re: node. I'd rather use the name `node` as the alternative `node_name` is kinda meh. also we aren't really dealing with distribution here, so `node` even if shadowed is kinda irrelevant.

Reply:
really, we may want to call this `members` or something. but that means breaking compatibility with the existing HashRing module.

Reply:
WRT spaces; I get that we do that before (so did Elixir core, which has moved away from this practice), but I find it *very* hard to read a giant block of unspaced code. One of the nice things about elixir is that the function heads are only a couple of lines, and this obviates that benefit. 

Also, last night, you criticized the old API for not being consistent and now want to maintain compatibility. You don't need to live and die by compatibility, if something is better, rev the version number and go for it.